File: Data Structure/Hash/Medium/LET_no49_Group Anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> mp;
        for (string s : strs)
            mp[bucketSort(s)].push_back(s);
        
        vector<vector<string>> res;
        for (auto p : mp)
            res.push_back(p.second);
        
        return res;
    }
    
    string bucketSort(string& s) {
        vector<int> counts(26, 0);
        for (char c : s)
            counts[c - 'a']++;
        
        string res;
        for (int i = 0; i < 26; i++)
            res += string(counts[i], 'a' + i);
        
        return res;
    }
};

File: Data Structure/Hash/Medium/LET_no49_Group Anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        record = collections.defaultdict(list)
        for s in strs:
            record[self.bucketSort(s)].append(s)
        
        return [lst for lst in record.values()]
        
    def bucketSort(self, s):
        counts = [0] * 26
        for c in s:
            counts[ord(c) - ord('a')] += 1
        res = []
        for i in range(26):
            res.append(chr(ord('a') + i) * counts[i])
        return ''.join(res)